src/base/bittorrent/resumedatastorage.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2015-2022  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "resumedatastorage.h"
+
+#include <utility>
+
+#include <QHash>
+#include <QMetaObject>
+#include <QMutexLocker>
+#include <QThread>
+
+BitTorrent::ResumeDataStorage::ResumeDataStorage(QObject *parent)
+    : QObject(parent)
+{
+}
+
+void BitTorrent::ResumeDataStorage::loadAll() const
+{
+    auto *loadingThread = QThread::create([this]()
+    {
+        doLoadAll();
+    });
+    connect(loadingThread, &QThread::finished, loadingThread, &QObject::deleteLater);
+    loadingThread->start();
+}
+
+QVector<BitTorrent::LoadedResumeData> BitTorrent::ResumeDataStorage::fetchLoadedResumeData() const
+{
+    QMutexLocker locker {&m_loadedResumeDataMutex};
+    return std::exchange(m_loadedResumeData, {});
```

Comment:
Based on https://stackoverflow.com/a/45814404
`m_loadedResumeData.clear()` might be faster than `m_loadedResumeData = {};`, and the latter might change memory location if I understand correctly.

Reply:
Currently everything is moved. Without `std::exchange` it will copy at least once.

Reply:
> Test case 2:

It is not so correct. You forgot to add `QVector<int> b = std::move(a);`. Otherwize the result is the same as when use `std::exchange()` but the code takes 3 lines instead of just 1.

Reply:
> the point is after `m_loadedResumeData.clear()`, `m_loadedResumeData` still uses the same memory address.

It doesn't using both `std::move` and `std::exchange`.
Only copy could preserve it but it is bad alternative, IMO.

Reply:
OK, I saw my error.

Another thing, can `m_loadedResumeData` have a `reserve()`? probably with an estimated value.

And if `reserve()` is added then `std::exchange` will reset the capacity and you'll need to reserve it again. Since Qvector is implicitly shared there should still be no copy but a detachment, however detachment don't reset the capacity.